File: python/enolib/reporters/text_reporter.py
from ..constants import BEGIN, END, HUMAN_INDEXING
from .reporter import DISPLAY, EMPHASIZE, INDICATE, OMISSION, QUESTION, Reporter

INDICATORS = {
    DISPLAY: ' ',
    EMPHASIZE: '>',
    INDICATE: '*',
    QUESTION: '?'
}

class TextReporter(Reporter):
    def __init__(self, context):
        super().__init__(context)

        gutter_header = context.messages.gutter_header.rjust(5)
        columns_header = f"  {gutter_header} | {context.messages.content_header}"

        source = f"-- {context.source} --\n\n" if context.source is not None else ''

        self._gutter_width = len(gutter_header) + 3
        self._header = f"{source}{columns_header}\n"

    def _line(self, line, tag):
        if tag == OMISSION:
            return f"{' ' * (self._gutter_width - 5)}..."

        number = str(line + HUMAN_INDEXING)
        instruction = self._index[line]

        if instruction is None:
            content = ''
        else:
            content = self._context.input[instruction['ranges']['line'][BEGIN]:instruction['ranges']['line'][END]]

        return f" {INDICATORS[tag]}{number.rjust(self._gutter_width - 3)} | {content}"

    def _print(self):
        snippet = '\n'.join(self._line(line, tag) for line, tag in enumerate(self._snippet) if tag is not None)

        return self._header + snippet

File: ruby/lib/enolib/reporters/text_reporter.rb
# frozen_string_literal: true

module Enolib
  class TextReporter < Reporter
    INDICATORS = {
      display: ' ',
      emphasize: '>',
      indicate: '*',
      question: '?'
    }.freeze

    def initialize(context)
      super(context)

      gutter_header = context.messages::GUTTER_HEADER.rjust(5)
      columns_header = "  #{gutter_header} | #{context.messages::CONTENT_HEADER}"

      source = context.source ? "-- #{source} -- \n\n" : ''

      @gutter_width = gutter_header.length + 3
      @header = "#{source}#{columns_header}\n"
    end

    private

    def print_line(line, tag)
      return "#{' ' * (@gutter_width - 5)}..." if tag == :omission

      number = (line + Enolib::HUMAN_INDEXING).to_s
      instruction = @index[line]

      content =
        if instruction
          @context.input[instruction[:ranges][:line][0]...instruction[:ranges][:line][1]]
        else
          ''
        end

      " #{INDICATORS[tag]}#{number.rjust(@gutter_width - 3)} | #{content}"
    end

    def print
      @header + @snippet.each_with_index.map { |tag, line| print_line(line, tag) if tag }.compact.join("\n")
    end
  end
end
